abi-versions/v0.2.1/README.md

```diff
@@ -0,0 +1,1807 @@
+
+Returned `status` value is:
+- `OK` on success.
+- `NOT_FOUND` when the requested `queue_id` was not found.
+- `EMPTY` when there is nothing to dequeue from the requested queue.
+- `INVALID_MEMORY_ACCESS` when `return_value_data`
+  and/or `return_value_size` point to invalid memory address.
+
+
+### `proxy_on_queue_ready`
+* params:
+  - `i32 (uint32_t) context_id`
+  - `i32 (uint32_t) queue_id`
+* returns:
+  - none
+
+Called when a new item is enqueued on the queue `queue_id`.
+
+
+## Metrics
+
+### `proxy_define_metric`
+
+* params:
+  - `i32 (`[`proxy_metric_type_t`]`) metric_type`
+  - `i32 (const char *) name_data`
+  - `i32 (size_t) name_size`
+  - `i32 (uint32_t *) return_metric_id`
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Defines a new metric of `metric_type` with a `name`.
+
+Its value can be modified and/or retrieved using
+[`proxy_record_metric`], [`proxy_increment_metric`]
+and/or [`proxy_get_metric`] with `return_metric_id`.
+
+Returned `status` value is:
+- `OK` on success.
+- `BAD_ARGUMENT` for unknown `metric_type`.
+- `INVALID_MEMORY_ACCESS` when `name_data`, `name_size`
+  and/or `return_metric_id` point to invalid memory address.
+
+
+### `proxy_record_metric`
+
+* params:
+  - `i32 (uint32_t) metric_id`
+  - `i64 (uint64_t) value`
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Sets metric `metric_id` to the `value`.
+
+Returned `status` value is:
+- `OK` on success.
+- `NOT_FOUND` when the requested `metric_id` was not found.
+
+
+### `proxy_increment_metric`
+
+* params:
+  - `i32 (uint32_t) metric_id`
+  - `i64 (int64_t) offset`
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Changes metric `metric_id` by the `offset`.
+
+Returned `status` value is:
+- `OK` on success.
+- `NOT_FOUND` when the requested `metric_id` was not found.
+- `BAD_ARGUMENT` when requested `offset` cannot be applied to
+  `metric_id` (e.g. trying to decrement counter).
+
+
+### `proxy_get_metric`
+
+* params:
+  - `i32 (uint32_t) metric_id`
+  - `i32 (uint64_t *) return_value`
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Retrieves `return_value` of the metric `metric_id`.
+
+Returned `status` value is:
+- `OK` on success.
+- `NOT_FOUND` when the requested `metric_id` was not found.
+- `INVALID_MEMORY_ACCESS` when `return_value` points to invalid memory
+  address.
+
+
+## Properties
+
+> **Warning**
+> Properties are implementation-dependent and not stable.
+
+### `proxy_get_property`
+
+* params:
+  - `i32 (const uint8_t *) path_data`
+  - `i32 (size_t) path_size`
+  - `i32 (uint8_t **) return_value_data`
+  - `i32 (size_t *) return_value_size`
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Returned `status` value is:
+- `OK` on success.
+- `NOT_FOUND` when there was no property found at the requested `path`.
+- `SERIALIZATION_FAILURE` TODO
+- `INVALID_MEMORY_ACCESS` when `path_data`, `path_size`,
+  `return_value_data` and/or `return_value_size` point to invalid
+  memory address.
```

Comment:
This is incomplete without a description of the encoding of the input / output I think.

Reply:
Honestly, I'm treating `proxy_{get,set}_property` in v0.2.1 as a opaque pass-thru to CEL (hence the comment at the beginning of this section), and I was hoping not to document existing paths, since the set isn't stable across  
host implementations.

Maybe we could redirect users to the documentation in proxies for CEL attributes?

In any case, I think CEL makes sense for a lot of use-cases (e.g. when used in expressions, or to access unspecified objects like Envoy's filter state or dynamic metadata), but I don't think it works great as a generic property getter for predefined values.

Notably:
1. Querying using strings vs enums is wasteful, and inconsistent with rest of the Proxy-Wasm ABI.
2. The set isn't stable, so it's impossible to enforce it at the ABI layer, and we had issues with that in the past.
3. There is a some overlap with existing functions (e.g. access to header maps), which is source of some confusion.

Reply:
> I don't think it works great as a generic property getter for predefined values.

huge +1